### DISCUSSION TITLE
RPi python

### DISCUSSION THREAD

**Discussion by clint70** [Curtis2012/dewheater#6] [Q&A]
Hi Curtis, looking at the dewheater code on my pi, seems that it's still running ver. 2.7.  what is the best way to upgrade this without breaking it?  I know I messed with the initial code to get it to work, but it's been stable and I don't feel like breaking it ;)

Also, this line in the status, what did I do to mess it up?  I think that is the section of code that has an if / else statement to show ON / OFF.

Thanks again for this great work... Love it.  I have the 2 bme's running.

**Comment by Curtis2012** [comment]
Do you mean running Python 2.7?

**Reply by clint70** [reply]
I want to run on python3, but right now, for whatever reason... I get errors

**Reply by Curtis2012** [reply]
Interesting. Im mostly off grid w no access to my machines for at least a week. Please post the errors you are getting and I will look at them when I can. 


For now, no harm in running 2.7.

**Reply by Curtis2012** [reply]
Re the " line in the status". I dont see any line posted in your message (poor internet here). Please post the line in reply.

**Reply by clint70** [reply]
Hi Curtis, the line pointed to a bad tab..  I corrected that.  Still won't run on my RPi 4 with the python3.  before I upgrade python, I want to make sure I don't break everything. :D Dewheater / Allsky etc.

I do have a question to follow up on, my status has a spurious 'new line' in it.  I was trying to figure out what I broke, as when I try to fix it, throws an error about print. :(

**Reply by clint70** [reply]
This may be dumb question, is there a log file, to show if / when the heater is on and off....  I don't think the dewheater is working for me?  BUT no way for me to see.

I do see DEW on the inside of the dome, been fight that for a few days now.  SOOOO my guess is that the dew heater is NOT working for me.  Might have messed up something.  BUT if I can find a log file, it'll point me at a starting place for troubleshooting.

**Reply by Curtis2012** [reply]
Clint, Python print formatting is funky and changes across versions. So post the suspect code here and I will take a look.

**Reply by Curtis2012** [reply]
Re log file. 

First run "dewheateroff.py" to turn that dew heater off! 

I plan to add a "maxOnTime" confirguration parameter to avoid that issue.o

The "journalctl" command can be used to view output of services. The following line should show you continuous output from dewheater.

journalctl -u dewheater.service -f

**Comment by clint70** [comment]
Thank you Curtis! Really appreciate it…

**Reply by Curtis2012** [reply]
Clint, re the dew heater staying on. Ive had this issue during testing when I forgot about temp differences between sensor location and under the dome. This is one reason that offsets are configurable. Its also why I plan to implememt a "maxOnTime" parameter.

Where is your sensor installed (under dome, in housing, etc)?

Please post your config file.

**Reply by clint70** [reply]
Morning Curtis, I think I may have miss led you, my concern is not staying on all the time, it is the opposite.  I don't think it comes on at all.  I have had nothing but troubles fighting dew on the inside of the dome. :(  When I was thinking the dewheater should prevent that.  

The sensor (BME280 is below the camera inside the housing, WHICH may be the issue?  OR not enough air flow?  BUT either way, I'll stick with the dewheater setup, but looking to replace the enclosure, one of the ppl on the Allsky discussion offered this one as a solution: https://www.thingiverse.com/thing:4721653

<img width="983" alt="image" src="https://user-images.githubusercontent.com/8769072/187034148-e98866c8-fbb4-4a27-9919-4a594c27065a.png">

 looks like the dewheater never turns on.... ?  

Config attached, along with the dewheater.py I'm using

[dewheater_files.zip](https://github.com/Curtis2012/dewheater/files/9438127/dewheater_files.zip)

**Reply by Curtis2012** [reply]
Clint, the temp at the location of your sensor, as shown in the log above, is 38C! Quite warm! This rising heat could be fogging your dome.

I have my sensor mounted under the dome.

Reads like yours is inside the housing w the Rpi so the temp reading will be affected by the Rpi and any other hardware in there. My housing is vented on the bottom to allow some passive ventilation. I have a cooling fan add on in the works.

For dew heater control you will be better off reading either temp under the dome or outside ambient temp.

**Reply by clint70** [reply]
Yes... I agree, the enclosure is small and the dome is like a greenhouse, heating up!!  That has been another issue I have been trying to solve.  Lots for me to learn and try to understand... :)

I have opted to move the BME from in the encloser to under the dome, as well as add a small fan to try and move the air around... this is as of today, tonight / tomorrow will tell if I have helped or not for the dew inside the dome.  I am thinking I am not that lucky :)  BUT time will tell.  In the end there are several factors... 
1. print the enclosure - better, so there is sealed walls
2. enlarge the enclosure
3. **maybe** add vent holes on the bottom, but I'm not sold on this, as it will allow 'critters' to get in. (I hate spiders ;) ).
![IMG_0448](https://user-images.githubusercontent.com/8769072/187099169-d71408b2-2640-4612-ac13-84edc8fd96e7.jpeg)
![IMG_0449](https://user-images.githubusercontent.com/8769072/187099170-beca947c-60d1-4e33-8dda-c8708dbf25eb.jpeg)

**Reply by Curtis2012** [reply]
That looks like a move in the right direction! 

A bit of fine mesh or a scotch-brite pad cut to fit allows air flow, but keeps critters out.

I sprayed my dome white except for right over the lens. Helped cut the heat.

Upgrade Phython! There really are not many compatibilty issues.

**Reply by clint70** [reply]
Hahaha, yes i know, ;). I'll upgrade this week, first look after stability. I like your idea about painting the dome white, I'll have to find out where the edge's of the image frame is :). But not sure how to do that just yet.  

Thanks for the advice.

**Reply by clint70** [reply]
Ok... so I will apologize up front, as this may be a 'dumb' question, but I can't see the dewheater ever turning on.  I looked at the code, and not sure I see how the conditions are met for it to turn on... right now this is my status:

```Aug 28 23:53:47 lr-allsky.cyberrover.ca python[364]: Dew point met = False, fakeDewPoint = False, fakeDewPointCounter = 0
Aug 28 23:53:47 lr-allsky.cyberrover.ca python[364]: ====================================================
Aug 28 23:53:52 lr-allsky.cyberrover.ca python[364]: ====================================================
Aug 28 23:53:52 lr-allsky.cyberrover.ca python[364]: Temp = 36.4C, temp_actual = 36.4C, Humidity 35.6% Dew Point = 18.7C
Aug 28 23:53:52 lr-allsky.cyberrover.ca python[364]: Dew heater state =  OFF
Aug 28 23:53:52 lr-allsky.cyberrover.ca python[364]: invertOnOff = True
Aug 28 23:53:52 lr-allsky.cyberrover.ca python[364]: MinTempOn set point = 1.0C, MinTempOn = False
Aug 28 23:53:52 lr-allsky.cyberrover.ca python[364]: MaxTempOff set point = 35.0C, MaxTempOff = True
Aug 28 23:53:52 lr-allsky.cyberrover.ca python[364]: Dew point met = False, fakeDewPoint = False, fakeDewPointCounter = 0
Aug 28 23:53:52 lr-allsky.cyberrover.ca python[364]: ====================================================
```
The current temp in the dome reads: 

```pi@lr-allsky:~ $ python Downloads/bme280.py
Chip ID     : 96
Version     : 0
Temperature :  36.15 C
Pressure :  1002.41221486 hPa
Humidity :  35.7859397015 %
```

Right now there is no dew in the dome... SCORE... and there is a ton of humidity.  Looking to understand why I never see the dewheater meet it's conditions?  Maybe that is part of my issue, OR because the enclosure is so warm inside, that the dewheater never meets its conditions??  SORRY for the trouble...

**Reply by clint70** [reply]
Not sure about the dewheater... but this AM, no moisture inside the dome.  So that is a move in the right direction.  Might be the fan moving the air around??  IDK

**Reply by Curtis2012** [reply]
Clint,  the temp at your sensor per log is 36.4C (toasty!). The calculated dew point is 18.7C.  The dew heater will not turn on until actual temp is equal to or less than the dew point.

My guess is that at such warm temps inside your dome/housing, you are effectively dehumidifying the air in the enclosure and the warm moisture has nowhere to go except up where it condenses against the dome...without ever reaching dew point conditions inside the dome. 

The fan probably helped with this by keeping the air mixed.

**Reply by Curtis2012** [reply]
Humidity at your sensor is 35.7% (low). What is the ambient humidity (outside)?

**Reply by clint70** [reply]
Current 'ambient' humididy is 65% it's muggy right now.  STILL no moisture on the interior of the dome, so perhaps you are right.  The fan moving the air has successfully helped (for the moment)  Although still gets crazy hot  inside.  I am thinking your idea of painting out the dome white... is good.  As well, I may re-print the enclosure in ASA and put a ventilation hole on the bottom (?) we'll see.

Ok, so if I understand, potentially the dewheater may never come on... because the enclosure is so warm.  FOR now... but maybe when winter hits (here in Canada) it will be more effective.

**Reply by Curtis2012** [reply]
Yep, you have built a dehumidifyer with nowhere for the moisture to go!  The internal humdity is about 35%, external (ambient) is about 65%...a BIG drop!

Those are also hot temps for the RaspberryPi, it is likely throttling the CPU to try and reduce temp.

The fan helps keep the air mixed so reduces the condensation problem. 

Adding some ventilation should help both the moisture and heat issues.

Correct, the dew heater will never come on becasue dew point is never being reached.

Note: if you just want to see if your dew heater relay is switching on/off then use the "fakeDewPoint" feature to test.

**Reply by clint70** [reply]
OK... so, bummed - big storm rolled through, pretty sure the enclosure is NOT water tight UGH!  I think since it's not water tight, a bit of spray flex seal on the outside, I'll drill the holes - fill with fine steel-wool.   AND the fans / dewheater should work 'better' and there should be an escape for moisture??  

I do think a new printed enclosure is crucial at this point.  BUT I don't want to have the Allsky down :(

**Reply by clint70** [reply]
Ok... so moisture flooded the enclosure, 70% relative humidity :O... dome covered in moisture inside.  So I am playing with the 'fake dewpoint'.  

```pi@lr-allsky:~ $ systemctl status dewheater.service 
● dewheater.service - dewheater
   Loaded: loaded (/lib/systemd/system/dewheater.service; enabled; vendor preset: enabled)
   Active: active (running) since Mon 2022-08-29 22:38:23 EDT; 10min ago
 Main PID: 10621 (python)
    Tasks: 1 (limit: 3596)
   CGroup: /system.slice/dewheater.service
           └─10621 /usr/bin/python /home/pi/dewheater/dewheater.py

Aug 29 22:48:45 lr-allsky.cyberrover.ca python[10621]: Dew point met = True, fakeDewPoint = 38, fakeDewPointCounter = 124
Aug 29 22:48:45 lr-allsky.cyberrover.ca python[10621]: ====================================================
Aug 29 22:48:50 lr-allsky.cyberrover.ca python[10621]: ====================================================
Aug 29 22:48:50 lr-allsky.cyberrover.ca python[10621]: Temp = 27.9C, temp_actual = 37.7C, Humidity 64.8% Dew Point = 29.9C
Aug 29 22:48:50 lr-allsky.cyberrover.ca python[10621]: Dew heater state =  ON
Aug 29 22:48:50 lr-allsky.cyberrover.ca python[10621]: invertOnOff = False
Aug 29 22:48:50 lr-allsky.cyberrover.ca python[10621]: MinTempOn set point = 1.0C, MinTempOn = False
Aug 29 22:48:50 lr-allsky.cyberrover.ca python[10621]: MaxTempOff set point = 40.0C, MaxTempOff = False
Aug 29 22:48:50 lr-allsky.cyberrover.ca python[10621]: Dew point met = True, fakeDewPoint = 38, fakeDewPointCounter = 125
Aug 29 22:48:50 lr-allsky.cyberrover.ca python[10621]: ====================================================
```

the teme shows 27.9 and dew point is 29.9... should the dewheater not be on?  It didn't turn on, I had to force it. Either way, when the heater is on, it is slowly clearing out the moisture.  So that is awesome, just need to figure out what I am doing wrong ;) 

<img width="1098" alt="image" src="https://user-images.githubusercontent.com/8769072/187338300-3c3a0fa3-6665-4383-8110-2801dc9b5150.png">

**Reply by Curtis2012** [reply]
Yes, the dew heater should be on and the log shows that its status is on. 

Can you force it on/off  with dewheateron.py and dewheateroff.py?

**Reply by Curtis2012** [reply]
Bummer about the rain. Thats a big print. Why not just use PVC pipe and print an adapter flange for it? Way shorter print time and PVC is waterproof for sure.

**Reply by clint70** [reply]
> Yes, the dew heater should be on and the log shows that its status is on.
> 
> Can you force it on/off with dewheateron.py and dewheateroff.py?

No, or at least when I 'force' it on with the dewheateron.py, it does not indicate on the status.  IDK why, will investigate.

**Reply by clint70** [reply]
> Bummer about the rain. Thats a big print. Why not just use PVC pipe and print an adapter flange for it? Way shorter print time and PVC is waterproof for sure.

Yes.. PVC would be the way to go, but RN everything is so $$ and we can't buy anything less than 4' section.  That 4' section is $35.00 CAN.  I can print a lot for 35.00 ;)  Mind you if they turn out like the last one... hahaha might not be worth it.

ATM, spray sealed the outside of the enclosure, and forcing the dewheater on all night dried everything out so far.  SOOOO IDK, might be ok for the moment.  BUT the last status was:

```● dewheater.service - dewheater
   Loaded: loaded (/lib/systemd/system/dewheater.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2022-08-30 13:30:01 EDT; 7s ago
 Main PID: 29701 (python)
    Tasks: 1 (limit: 3596)
   CGroup: /system.slice/dewheater.service
           └─29701 /usr/bin/python /home/pi/dewheater/dewheater.py

Aug 30 13:30:01 lr-allsky.cyberrover.ca python[29701]: /home/pi/dewheater/dewheater.py:79: RuntimeWarning: This channel is already in use, continuing anyway.  Use
Aug 30 13:30:01 lr-allsky.cyberrover.ca python[29701]:   GPIO.setup(self.dewHeaterPin, GPIO.OUT)
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: ====================================================
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: Temp = 40.8C, temp_actual = 40.8C, Humidity 56.0% Dew Point = 30.3C
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: Dew heater state =  OFF
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: invertOnOff = False
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: MinTempOn set point = 1.0C, MinTempOn = False
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: MaxTempOff set point = 40.0C, MaxTempOff = True
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: Dew point met = False, fakeDewPoint = False, fakeDewPointCounter = 0
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: ====================================================
```
It is super humid right now... and WARM... now the enclosure is 'hot...ish' but the RPi CPU is not running hot, so ok for now.  BUT once direct sun hits it, it will prob sky rocket.

**Reply by clint70** [reply]
So... I don't know... just seems like the BME readings are off, I can't get the dewheater to come on, by itself.  I can use the fake dewpoint, then force on for a long time... which does work... removes the moisture from the inside of the dome, but... ??  IDK how to get this to do it 'automatically'.

Moisture in the dome RN, 

```● dewheater.service - dewheater
   Loaded: loaded (/lib/systemd/system/dewheater.service; enabled; vendor preset: enabled)
   Active: active (running) since Tue 2022-08-30 13:30:01 EDT; 7s ago
 Main PID: 29701 (python)
    Tasks: 1 (limit: 3596)
   CGroup: /system.slice/dewheater.service
           └─29701 /usr/bin/python /home/pi/dewheater/dewheater.py

Aug 30 13:30:01 lr-allsky.cyberrover.ca python[29701]: /home/pi/dewheater/dewheater.py:79: RuntimeWarning: This channel is already in use, continuing anyway.  Use
Aug 30 13:30:01 lr-allsky.cyberrover.ca python[29701]:   GPIO.setup(self.dewHeaterPin, GPIO.OUT)
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: ====================================================
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: Temp = 40.8C, temp_actual = 40.8C, Humidity 56.0% Dew Point = 30.3C
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: Dew heater state =  OFF
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: invertOnOff = False
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: MinTempOn set point = 1.0C, MinTempOn = False
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: MaxTempOff set point = 40.0C, MaxTempOff = True
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: Dew point met = False, fakeDewPoint = False, fakeDewPointCounter = 0
Aug 30 13:30:04 lr-allsky.cyberrover.ca python[29701]: ====================================================
```

Forced on by fake dewpoint, and we'll see if that clears up the dome again, it worked well last night doing that, but I had to force like 1000 cycles.

**Reply by Curtis2012** [reply]
If the dew point is never met then it will never come on automatically. So far in the logs you have posted, Ive never seen the dew point met.

You can force it on with dewheateron.py.

**Reply by Curtis2012** [reply]
> > Yes, the dew heater should be on and the log shows that its status is on.
> > 
> > Can you force it on/off with dewheateron.py and dewheateroff.py?
> 
> No, or at least when I 'force' it on with the dewheateron.py, it does not indicate on the status.  IDK why, will investigate.

It wont show in the status from dewheater service, but it should close the relay. 

Best to stop the service first then force on with dewheateron.py. Later turn off with dewheateroff.py.  

These 2 utilities are just for testing so they do not cause the service to update its state of the dew heater relay.

You can use them to test that your dewheater is switching on and off. Or just to force it on/off.

**Comment by clint70** [comment]
Here is the code causing me issue ;)

```def dispalySatus():
    print("====================================================")
    print("Temp = %3.1fC, temp_actual = %3.1fC, Humidity %3.1f%% Dew Point = %3.1fC" % (
        conditions.temperature, conditions.temp_actual, conditions.humidity, conditions.dewPoint.c))
    #print("Dew heater state =", end= " ")
    print("Dew heater state = ")
    if (dewHeater.status == ON):
        print("ON")
    else:
        print("OFF")
    print("invertOnOff = %s" % (config.invertOnOff))
    print("MinTempOn set point = %3.1fC, MinTempOn = %s" % (config.dewHeaterMinTemp, dewHeater.minTempOn))
    print("MaxTempOff set point = %3.1fC, MaxTempOff = %s" % (config.dewHeaterMaxTemp, dewHeater.maxTempOff))
    print("Dew point met = %s, fakeDewPoint = %s, fakeDewPointCounter = %i " % (conditions.dewPointMet, config.fakeDewPoint, conditions.fakeDewPointCounter))
    print("====================================================")

```

The output looks like this:

```Aug 27 00:07:05 lr-allsky.cyberrover.ca python[386]: ====================================================
Aug 27 00:07:10 lr-allsky.cyberrover.ca python[386]: ====================================================
Aug 27 00:07:10 lr-allsky.cyberrover.ca python[386]: Temp = 28.8C, temp_actual = 28.8C, Humidity 57.2% Dew Point = 19.5C
Aug 27 00:07:10 lr-allsky.cyberrover.ca python[386]: Dew heater state =
Aug 27 00:07:10 lr-allsky.cyberrover.ca python[386]: OFF
Aug 27 00:07:10 lr-allsky.cyberrover.ca python[386]: invertOnOff = False
Aug 27 00:07:10 lr-allsky.cyberrover.ca python[386]: MinTempOn set point = 1.0C, MinTempOn = False
Aug 27 00:07:10 lr-allsky.cyberrover.ca python[386]: MaxTempOff set point = 40.0C, MaxTempOff = False
Aug 27 00:07:10 lr-allsky.cyberrover.ca python[386]: Dew point met = False, fakeDewPoint = False, fakeDewPointCounter = 0
Aug 27 00:07:10 lr-allsky.cyberrover.ca python[386]: ====================================================```

AND I don't ever see the dewheater coming on... :O

**Reply by Curtis2012** [reply]
By default the "print" statement outputs a carriage return (new line). You can change that by specifying a different ending character with the "end=" paramter. 

See the current dewheater.py in the "merge_BME280_and_DHT_support" branch for an example...or just run this code.

**Reply by clint70** [reply]
Curtis, nope... doesn't work for python 2.X.  The solution was 2.7 specific, by adding the 'comma' at the end of the line you don't want the 'new line character' which fixed my issue.  I know it's specific to my code / version 2.x python.  I will look at upgrading to stop me chasing all these old compatibility issues.

Thank you for the time / help.  I do enjoy this... and your coding and work are very helpful.  Output now:

```● dewheater.service - dewheater
   Loaded: loaded (/lib/systemd/system/dewheater.service; enabled; vendor preset: enabled)
   Active: active (running) since Sun 2022-08-28 15:09:59 EDT; 3h 35min ago
 Main PID: 359 (python)
    Tasks: 1 (limit: 3596)
   CGroup: /system.slice/dewheater.service
           └─359 /usr/bin/python /home/pi/dewheater/dewheater.py

Aug 28 18:45:46 lr-allsky.cyberrover.ca python[359]: Dew point met = False, fakeDewPoint = False, fakeDewPointCounter = 0
Aug 28 18:45:46 lr-allsky.cyberrover.ca python[359]: ====================================================
Aug 28 18:45:51 lr-allsky.cyberrover.ca python[359]: ====================================================
Aug 28 18:45:51 lr-allsky.cyberrover.ca python[359]: Temp = 41.2C, temp_actual = 41.2C, Humidity 27.0% Dew Point = 18.5C
Aug 28 18:45:51 lr-allsky.cyberrover.ca python[359]: Dew heater state =  OFF
Aug 28 18:45:51 lr-allsky.cyberrover.ca python[359]: invertOnOff = True
Aug 28 18:45:51 lr-allsky.cyberrover.ca python[359]: MinTempOn set point = 1.0C, MinTempOn = False
Aug 28 18:45:51 lr-allsky.cyberrover.ca python[359]: MaxTempOff set point = 35.0C, MaxTempOff = True
Aug 28 18:45:51 lr-allsky.cyberrover.ca python[359]: Dew point met = False, fakeDewPoint = False, fakeDewPointCounter = 0
Aug 28 18:45:51 lr-allsky.cyberrover.ca python[359]: ====================================================
```

**Comment by clint70** [comment]
I think this fixes it for python 2.X

```def dispalySatus():
    print("====================================================")
    print("Temp = %3.1fC, temp_actual = %3.1fC, Humidity %3.1f%% Dew Point = %3.1fC" % (
        conditions.temperature, conditions.temp_actual, conditions.humidity, conditions.dewPoint.c))
    print("Dew heater state = "),
    if (dewHeater.status == ON):
        print("ON")
    else:
        print("OFF")
    print("invertOnOff = %s" % (config.invertOnOff))
    print("MinTempOn set point = %3.1fC, MinTempOn = %s" % (config.dewHeaterMinTemp, dewHeater.minTempOn))
    print("MaxTempOff set point = %3.1fC, MaxTempOff = %s" % (config.dewHeaterMaxTemp, dewHeater.maxTempOff))
    print("Dew point met = %s, fakeDewPoint = %s, fakeDewPointCounter = %i " % (conditions.dewPointMet, config.fakeDewPoint, conditions.fakeDewPointCounter))
    print("====================================================")```

**Comment by clint70** [comment]
Curtis, so been a week or so, the enclosure is dry... the heater does come on, I find the library calculation for the dew point to be close, but because the enclosure tends to have more moisture *ATM* I am forcing the heater to come on more often by adding to the dew point.  This has worked well for me.

I added a section to the dewheaterconfig.json (offset) so I can easily update my dewpoint.  Probably a 'terrible' idea, but I needed some way to force the heater on more often :P 

```{
  "debug": true,
  "dewHeaterPin": 23,
  "dewHeaterCutinOffset": 1.0,
  "dewHeaterCutoutOffset": 1.0,
  "dewHeaterMaxTemp": 40.0,
  "dewHeaterMinTemp": 1.0,
  "dewHeaterOnOffDelay": 5,
  "dewPtCheckDelay": 5,
  "fakeDewPoint": false,
  "fakeDewPointSamples": 2500,
  "invertOnOff": false,
  "offset": 9
}

```

<img width="914" alt="image" src="https://user-images.githubusercontent.com/8769072/188891381-96aa9450-41c4-4c24-a165-d83bac5dab6c.png">

**Reply by Curtis2012** [reply]
Congratulations!

Ha,  its not a terrible idea...in fact its already in the code! 

The two fields below allow you to offset either the "cut-in" or "cut-out" temps by an offset factor. The "cut-in" temp is the temp at which the dew heater relay is closed (on). The "cut-out" temp is the temp at which it is opened (off). Either can be set to an integer value (+/-).

"dewHeaterCutinOffset": 1.0,
"dewHeaterCutoutOffset": 1.0,

**Reply by clint70** [reply]
DOH! missed that, guess I didn't really understand your code as well as I thought.  Sorry.  I will go play some more.  BUT again, thank you for the time/effort and putting this together.  I know your design was for the relay and camera dewheater.  I don't know if I mentioned that I have gone a different route.  I am using a setup with a mosfet that passes 12v to a ring of resistors (8x 6.5 ohms) that will heat up to 8 watts.  It works well with your code... but I don't have a relay in my setup. 

![image](https://user-images.githubusercontent.com/8769072/188914313-470e0cf4-1e9b-49dc-b6d5-a47749e486cb.png)

Here is what I currently have been getting with my site:  

https://lr-allsky.cyberrover.ca/allsky

**Reply by Curtis2012** [reply]
Cool. Should work just fine,  just another type of switch. Could be more reliable than some mechanical relays (most eventually have problems...sometime years though).

**Reply by Curtis2012** [reply]
The web site link times out for me.
